### ISSUE TITLE
Can this be used for planning in outdoor "3D" terrain?

### ISSUE THREAD

**Issue by jorgemia** [cogsys-tuebingen/gerona#28]
Can this package be used with a 3D map or elevation map to plan paths for a skid steer robot in outdoor non-flat terrain? Are there any examples?

**Comment by betwo** [comment]
Hi @jorgemia,

navigation in 2.5d is definitely possible, see for example [this video here](https://www.youtube.com/watch?v=4O7twdWFm4s).

Planning in full 3d (multiple layers) is not in the scope of this project, however.
You could use another planner such as from OMPL and only use the control aspects of GeRoNa (might require some conversion code for different interfaces).

If you have an elevation / 2.5d map, you can write some glue-code and should be able to use the existing planners.
The planners we provide here assume that the world is flat, though, so there is no practical way to find *optimal* paths in 2.5d because all slope information is lost.

What you would need is a ROS node that extracts obstacle information from the elevation map. 
The simplest way might be to define a maximum slope and then sample the neighborhood around the robot.
We had something similar working in the past, but we never published it with this framework.

Here's a related example where we did something similar directly on the measurement of a 3d sensor:
https://github.com/cogsys-tuebingen/csapex_generic_obstacle_detection
This was used in the [video linked above](https://www.youtube.com/watch?v=4O7twdWFm4s).

**Comment by jorgemia** [comment]
Thanks for the answer @betwo !

Ok seems I might have to tweak a couple of things to make it work. For now I'm only worried about 2.5D nav but maybe in the future multi-level would be nice. I have seen some projects that use OMPL but not as complete as GeRoNa in terms of features available. 

> The simplest way might be to define a maximum slope and then sample the neighborhood around the robot.
> We had something similar working in the past, but we never published it with this framework.

Yes, I read a paper called "Real-time model based path planning for wheeled vehicles" which mentioned GeRoNa could be integrated with it, is that the one?

If I end up using GeRoNa I'll let you know, thanks!

**Comment by betwo** [comment]
> Yes, I read a paper called "Real-time model based path planning for wheeled vehicles" which mentioned GeRoNa could be integrated with it, is that the one?

Cool, that paper is written by @jquadrat. The code is available here in [tools/model_based_planner](https://github.com/cogsys-tuebingen/gerona/tree/master/tools/model_based_planner).
This work used a local elevation map, as a local planner. Adapting this to global maps might break the assumptions @jquadrat  made for runtime optimization. Maybe he can elaborate?

What I was thinking about above was our work for the [Space Bot Cup 2015](http://www.cogsys.cs.uni-tuebingen.de/forschung/DLR_SpaceBot_Cup_2015/blog/), where we used elevation maps for the global planning. We had a costmap generated from 2.5d elevation and thresholded that for obstacle detection:
![](http://www.cogsys.cs.uni-tuebingen.de/forschung/DLR_SpaceBot_Cup_2015/blog/costmap_crop.png)

> If I end up using GeRoNa I'll let you know, thanks!

Nice, thank you :+1:

**Comment by jquadrat** [comment]
Hi @jorgemia, as @betwo  already mentioned, the code from "Real-time model based path planning for wheeled vehicles" is included in Gerona and can be found here [/tools/model_based_planner](https://github.com/cogsys-tuebingen/gerona/tree/master/tools/model_based_planner). It is integrated as a controller [robotcontroller_modelbased.cpp](https://github.com/cogsys-tuebingen/gerona/blob/master/path_follower/src/controller/robotcontroller_modelbased.cpp) which directly outputs commands and as a local planner that outputs a path for a path following controller [local_planner_model.cpp](https://github.com/cogsys-tuebingen/gerona/blob/master/path_follower/src/local_planner/model_based/local_planner_model.cpp).

For the controller implementation I've added an example launch file [gazebo_example_summit_mbc.launch](https://github.com/cogsys-tuebingen/gerona/blob/master/gerona_examples/launch/gazebo_example_summit_mbc.launch). (The map does not really shows the planning capabilities, should be replaced with something more challenging)

For creating a elevation map from RGB-D sensor data the [localmap](https://github.com/cogsys-tuebingen/gerona/tree/master/tools/localmap) can be used. 


Right now it is designed to be used for high frequency local planning on an elevation map.

Basically the image-based model representation could also be used to perform larger scale planning (if a sufficiently detailed elevation map is available) when used with a proper planning strategy. The pose estimator functionality (getting the 3D pose for a 2d position on an image) is wrapped in the [PoseEstimator](https://github.com/cogsys-tuebingen/gerona/blob/master/tools/model_based_planner/include/model_based_planner/poseestimator.h) class.

**Comment by jorgemia** [comment]
Thanks @jquadrat for the thorough answer!! I'll check this stuff out!
